src/lib/tls/tls_extensions.cpp

```diff
@@ -356,6 +363,108 @@ std::vector<uint8_t> Application_Layer_Protocol_Notification::serialize(Connecti
    return buf;
 }
 
+std::string certificate_type_to_string(Certificate_Type type) {
+   switch(type) {
+      case Certificate_Type::X509:
+         return "X509";
+      case Certificate_Type::RawPublicKey:
+         return "RawPublicKey";
+   }
+
+   return "Unknown";
+}
+
+Certificate_Type certificate_type_from_string(const std::string& type_str) {
+   if(type_str == "X509") {
+      return Certificate_Type::X509;
+   } else if(type_str == "RawPublicKey") {
+      return Certificate_Type::RawPublicKey;
+   } else {
+      throw Decoding_Error("Unknown certificate type: " + type_str);
+   }
+}
+
+Certificate_Type_Base::Certificate_Type_Base(std::vector<Certificate_Type> supported_cert_types) :
+      m_certificate_types(std::move(supported_cert_types)), m_from(Connection_Side::Client) {
+   BOTAN_ARG_CHECK(!m_certificate_types.empty(), "at least one certificate type must be supported");
+}
+
+Client_Certificate_Type::Client_Certificate_Type(const Client_Certificate_Type& cct, const Policy& policy) :
+      Certificate_Type_Base(cct, policy.accepted_client_certificate_types()) {}
+
+Server_Certificate_Type::Server_Certificate_Type(const Server_Certificate_Type& sct, const Policy& policy) :
+      Certificate_Type_Base(sct, policy.accepted_server_certificate_types()) {}
+
+Certificate_Type_Base::Certificate_Type_Base(const Certificate_Type_Base& certificate_type_from_client,
+                                             const std::vector<Certificate_Type>& server_preference) :
+      m_from(Connection_Side::Server) {
+   // RFC 7250 4.2
+   //    The server_certificate_type extension in the client hello indicates the
+   //    types of certificates the client is able to process when provided by
+   //    the server in a subsequent certificate payload. [...] With the
+   //    server_certificate_type extension in the server hello, the TLS server
+   //    indicates the certificate type carried in the Certificate payload.
+   for(const auto server_supported_cert_type : server_preference) {
+      if(value_exists(certificate_type_from_client.m_certificate_types, server_supported_cert_type)) {
+         m_certificate_types.push_back(server_supported_cert_type);
+         return;
+      }
+   }
+
+   // RFC 7250 4.2 (2.)
+   //    The server supports the extension defined in this document, but
+   //    it does not have any certificate type in common with the client.
+   //    Then, the server terminates the session with a fatal alert of
+   //    type "unsupported_certificate".
+   throw TLS_Exception(Alert::UnsupportedCertificate, "Failed to agree on certificate_type");
+}
+
+Certificate_Type_Base::Certificate_Type_Base(TLS_Data_Reader& reader, uint16_t extension_size, Connection_Side from) :
+      m_from(from) {
+   if(extension_size == 0) {
+      throw Decoding_Error("Certificate type extension cannot be empty");
+   }
+
+   if(from == Connection_Side::Client) {
+      const auto type_bytes = reader.get_tls_length_value(1);
```

Comment:
Do we need to check that the `extension_size` matches with how much we read here?

Reply:
Certainly doesn't hurt, thanks.